cartoframes/viz/kuviz.py

```diff
@@ -22,48 +31,72 @@ def all(credentials=None):
         kuvizs = km.all()
         return [kuviz_to_dict(kuviz) for kuviz in kuvizs]
 
-    def set_credentials(self, credentials=None):
-        self._credentials = credentials or get_default_credentials()
-        self._auth_client = _create_auth_client(self._credentials)
-
-    def publish(self, html, name, password=None):
-        return _create_kuviz(html=html, name=name, auth_client=self._auth_client, password=password)
-
-    def is_sync(self):
-        return all(layer.source.dataset.is_remote() for layer in self._layers)
+    def get_layers(self):
+        return self._layers
 
-    def is_public(self):
-        return all(layer.source.dataset.is_public() for layer in self._layers)
+    def set_layers(self, layers, name, table_name):
+        self._sync_layers(layers, table_name)
+        self._manage_maps_api_key(name)
+        self._add_layers_credentials()
 
-    def get_layers(self, maps_api_key='default_public'):
+    def publish(self, html, name, password=None):
+        self.kuviz = _create_kuviz(html=html, name=name, auth_client=self._auth_client, password=password)
+        return kuviz_to_dict(self.kuviz)
+
+    def update(self, data, name, password):
+        if not self.kuviz:
+            raise CartoException('The map has not been published. Use the `publish` method.')
+
+        self.kuviz.data = data
+        self.kuviz.name = name
+        self.kuviz.password = password
+        self.kuviz.save()
+
+        return kuviz_to_dict(self.kuviz)
+
+    def delete(self):
+        if self.kuviz:
+            self.kuviz.delete()
+            warn("Publication '{n}' ({id}) deleted".format(n=self._publisher.kuviz.name, id=self._publisher.kuviz.id))
+            self.kuviz = None
+            return True
+        return False
+
+    def _sync_layers(self, layers, table_name):
+        for idx, layer in enumerate(layers):
+            if layer.source.dataset.is_local():
+                table_name = normalize_name("{name}_{idx}".format(name=table_name, idx=idx))
+                layer = self._sync_layer(layer, table_name)
+            self._layers.append(layer)
+
+    def _sync_layer(self, layer, table_name):
+        layer.source.dataset.upload(table_name=table_name, credentials=self._credentials)
```

Comment:
Yes, if the map has more than 3 local layers, we will hit the limits (the current limit for that is 3, the practical one is bigger). It does not seem like an usual case, but it possible. 

On the other hand, we implement rate limit retries in the download, but we did not do it the upload. I am going to create an issue for that

Reply:
https://github.com/CartoDB/cartoframes/issues/1095

Reply:
I think we should revisit the COPY limits instead of working on retry stuff.

I have a big concern about users hitting rate limits silently and blaming about this (CARTOframes) not working at all.

I'm adding a comment to that issue, to see if we can have a conversation about this next week.